### PR TITLE
Add Reference "component"

### LINKED ISSUE #210: Adding component references
In IDAES, we frequently have situations where we would like to reference a Pyomo component (most frequently a Param or Set) from one Block in another Block within our models. The most common case for this is where we have a Param that is used in multiple sub-Blocks within a model, but would like have a single, centralised Param object (so that the parameter only needs to be changed in a single location, rather than in every instance).

At the moment I do this using weakrefs to create the necessary links, which requires me to dereference the weakref whenever I use it. However, this results in some objects in my constraints having () after them and some without, which will make it difficult for new users to understand my code.

It was suggested that a component reference feature could be built into Pyomo to handle this transparently, without the need for the ().

### LINKED ISSUE #227: Connectors do not support sliced variables
I am trying to make a Connector linking between two sliced variables. See example below, along with stack trace.

```
from pyomo.environ import *
from pyomo.core.base.connector import ConnectorExpander
m = ConcreteModel()
# Testing sliced variables in connectors
m.s = Set(initialize=['A', 'B'])
m.s2 = Set(initialize=[1, 2])
m.s3 = Set(initialize=[1, 2, 3])
m.b1 = Block()
m.b1.x = Var(m.s, m.s2, initialize=1)
m.c = Connector()
m.c.add(m.b1.x[:, 2])

m.b2 = Block()
m.b2.x = Var(m.s, m.s3, initialize=2)
m.c2 = Connector()
m.c2.add(m.b2.x[:, 1])
m.constr = Constraint(expr=m.c == m.c2)

xfrm = ConnectorExpander()
xfrm.apply(instance=m)
```

```
AttributeError                            Traceback (most recent call last)
<ipython-input-9-51347db8734a> in <module>()
     15 
     16 xfrm = ConnectorExpander()
---> 17 xfrm.apply(instance=m)

/home/qichen/.solvers/pyomo/src/pyomo/pyomo/core/base/connector.pyc in apply(self, **kwds)
    282         instance = kwds.pop('instance')
    283         xform = TransformationFactory('core.expand_connectors')
--> 284         xform.apply_to(instance, **kwds)
    285         return instance
    286 

/home/qichen/.solvers/pyomo/src/pyomo/pyomo/core/base/plugin.pyc in apply_to(self, model, **kwds)
    333         if not hasattr(model, '_transformation_data'):
    334             model._transformation_data = TransformationData()
--> 335         self._apply_to(model, **kwds)
    336 
    337     def create_using(self, model, **kwds):

/home/qichen/.solvers/pyomo/src/pyomo/pyomo/core/plugins/transform/expand_connectors.pyc in _apply_to(self, instance, **kwds)
     89                 continue
     90             known_conn_sets[id(conn_set)] \
---> 91                 = self._validate_and_expand_connector_set(conn_set)
     92 
     93         # Expand each constraint

/home/qichen/.solvers/pyomo/src/pyomo/pyomo/core/plugins/transform/expand_connectors.pyc in _validate_and_expand_connector_set(self, connectors)
    143                     #-3 if v is None else
    144                     -2 if k in c.aggregators else
--> 145                     -1 if not hasattr(v, 'is_indexed') or not v.is_indexed()
    146                     else len(v) )
    147                 ref[k] = ( v, _len, c )

/home/qichen/.solvers/pyomo/src/pyomo/pyomo/core/base/indexed_component.pyc in __call__(self, *idx, **kwds)
    236             # Note: simply calling "list(self)" results in infinite
    237             # recursion in python2.6
--> 238             return list( i for i in self )
    239 
    240 

/home/qichen/.solvers/pyomo/src/pyomo/pyomo/core/base/indexed_component.pyc in <genexpr>(***failed resolving arguments***)
    236             # Note: simply calling "list(self)" results in infinite
    237             # recursion in python2.6
--> 238             return list( i for i in self )
    239 
    240 

/home/qichen/.solvers/pyomo/src/pyomo/pyomo/core/base/indexed_component.pyc in next(self)
     77     def next(self):
     78         """__next__() iterator for Py2 compatibility"""
---> 79         return self.__next__()
     80 
     81     def __next__(self):

/home/qichen/.solvers/pyomo/src/pyomo/pyomo/core/base/indexed_component.pyc in __next__(self)
    105                 if _call[0] == _IndexedComponent_slicer.attribute:
    106                     try:
--> 107                         _comp = getattr(_comp, _call[1])
    108                     except AttributeError:
    109                         # Since we are slicing, we may only be interested in

AttributeError: 'str' object has no attribute 'is_indexed'
```

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
This adds a "component `Reference` (actually a function) that takes a component slice and returns an indexed component whose data is dictated by that slice.  The actual data is never copied, so any changes to the components referenced by the slice will be reflected in the reference component.

## Changes proposed in this PR:
- Add `Reference` method for creating reference components
- Split the _IndexedComponent_slicer into separate `_IndexedComponent_slice` and `_IndexedComponent_slice_iter` classes, so that slices may be iterated over multiple times
- Move the slice code into a separate module

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/core/base/reference.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        >>> m.r3.pprint()
+        r3 : Size=4, Index=r3_index
+            Key    : Lower : Value : Upper : Fixed : Stale : Domain
+            (1, 3) :     1 :  None :  None : False :  True :  Reals
+            (1, 4) :     1 :  None :  None : False :  True :  Reals
+            (2, 3) :     2 :  None :  None : False :  True :  Reals
+            (2, 4) :     2 :  None :  None : False :  True :  Reals
+
+    The resulting reference component may be used just like any other
+    component.  Changes to the stored data will be reflected in the
```
Comment (blnicho) [inline pyomo/core/base/reference.py]: Does this include using them in Constraints?

Comment (jsiirola) [reply pyomo/core/base/reference.py]: Yes.  The data object you get back (e.g., by `m.r[2,4]`) *is* the object from the referenced components, so you can use them just like any other component.  Note though that since the Reference must be constructed from a component slice, the component returned by `Reference()` will always be indexed (and therefore `m.r` cannot be used in a constraint).

File: pyomo/core/base/indexed_component_slice.py [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+        the slice and call the item.  This allows "vector-like" operations
+        like: `m.x[:,1].fix(0)`.
+        """
+        self._call_stack.append( (
+            _IndexedComponent_slice.call, idx, kwds ) )
+        if self._call_stack[-2][1] == 'component':
+            return self
+        else:
+            # Note: simply calling "list(self)" results in infinite
+            # recursion in python2.6
```
Comment (blnicho) [inline pyomo/core/base/indexed_component_slice.py]: Didn't we drop support for Python 2.6?

File: pyomo/core/base/indexed_component_slice.py [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+                self.last_index = _idx
+                # Note: it is important to use __getitem__, as the
+                # derived class may implement a non-standard storage
+                # mechanism (e.g., Param)
+                return self.component[index]
+
+# Mock up a callable object with a "check_complete" method
+def _advance_iter(_iter):
+    return advance_iterator(_iter)
+def _advance_iter_check_complete():
```
Comment (blnicho) [inline pyomo/core/base/indexed_component_slice.py]: Why is this needed?

File: pyomo/core/base/indexed_component_slice.py [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+                        # so we don't need the overhead of the
+                        # _IndexedComponent_slice object)
+                        assert len(_comp._call_stack) == 1
+                        self._iter_stack[idx] = _slice_generator(
+                            *_comp._call_stack[0][1])
+                        try:
+                            _comp = self.advance_iter(self._iter_stack[idx])
+                        except StopIteration:
+                            # We got a slicer, but the slicer doesn't
+                            # matching anything.  We should break here,
```
Comment (blnicho) [inline pyomo/core/base/indexed_component_slice.py]: Typo: "doesn't matching anything"

File: pyomo/core/base/reference.py [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    :py:meth:`Component.type`, then that type is assumed.  If either the
+    ctype parameter is ``None`` or the data has more than one ctype, the
+    resulting indexed component will have a ctype of
+    :py:class:`IndexedComponent`.
+
+    If the indices associated with wildcards in the component slice all
+    refer to the same :py:class:`Set` objects for all data identifed by
+    the slice, then the resulting indexed component will be indexed by
+    the product of those sets.  However, if all data do not share common
+    set objects, or only a subset of indices in a multidimentional set
```
Comment (blnicho) [inline pyomo/core/base/reference.py]: Typo: "multidimentional"

File: pyomo/core/tests/unit/test_indexed.py [hunk 6]
```diff
@@ -18,9 +18,8 @@
 import pyutilib.th as unittest
 
 from pyomo.environ import *
-from pyomo.core.base.block import _BlockData
-from pyomo.core.base.indexed_component import _IndexedComponent_slicer
 
+from six import itervalues
```
Comment (blnicho) [inline pyomo/core/tests/unit/test_indexed.py]: You added an import that you never used in this file
